Project-Scoped Image Pull Secrets for Mirrored Registries: TP (#111160)
openshift/openshift-docs: modules/images-configuration-registry-mirror-project-secret.adoc

:_mod-docs-content-type: PROCEDURE
[id="images-configuration-registry-mirror-project-secret_{context}"]
= Configuring project-scoped image pull secrets for mirrored registries

[role="_abstract"]
As a cluster administrator, you can configure a `CRIOCredentialProviderConfig` object to enable project-scoped image pull secrets that you can use with mirrored repositories. By using project-scoped secrets, you can maintain security isolation between projects without exposing credentials in a global pull secret.

By default, if your cluster uses an `ImageDigestMirrorSet`, `ImageTagMirrorSet`, or `ImageContentSourcePolicy` object to configure repository mirroring, you must use a global pull secret for mirrored registries. You cannot add an image pull secret to a project. However, you can use a cluster-wide `CRIOCredentialProviderConfig` object to trigger the CRI-O credential provider, which enables project-scoped image pull secrets.

This feature is supported for clusters on the {aws-first}, {gcp-first}, or {azure-first} platform.

:FeatureName: Project-scoped image pull secrets for mirrored registries
include::snippets/technology-preview.adoc[]

> An administrator creates a cluster-wide `CRIOCredentialProviderConfig` object that lists the registries that a developer can pull from by using a using project-scoped secrets. When a developer uses a pod spec to pull an image from a listed registry, the `CRIOCredentialProviderConfig` object triggers the CRI-O credential provider. The credential provider resolves mirror configurations, discovers namespace-scoped secrets, and generates short-lived authentication files for CRI-O consumption. This process maintains credential isolation between namespaces while preserving existing mirror configuration methods.

Reviewer: "by using a using project-scoped secrets" has a duplicate word. I assume it should be "by using project-scoped secrets."

----
apiVersion: kubelet.config.k8s.io/v1
kind: CredentialProviderConfig
providers:
- apiVersion: credentialprovider.kubelet.k8s.io/v1
  args:
  - get-credentials
  - --v=3
  defaultCacheDuration: 1m0s
  matchImages:
  - gcr.io
  - '*.gcr.io'
  - '*.pkg.dev'
  - container.cloud.google.com
  name: gcr-credential-provider
- apiVersion: credentialprovider.kubelet.k8s.io/v1
  defaultCacheDuration: 1s
  matchImages:
  - '*.example.io'
  - docker.io
  - quay.io
  - registry.example.com:5000
  name: crio-credential-provider
  tokenAttributes:
    cacheType: Token
    requireServiceAccount: false
    serviceAccountTokenAudience: https://kubernetes.default.svc
----
where:

`providers.defaultCacheDuration`:: Specifies the default duration the plugin will cache credentials in-memory. The default is `1m0s`.
`providers.matchImages`:: Specifies the registries you listed in the `CRIOCredentialProviderConfig` object.
`providers.tokenAttributes`:: Specifies the configuration of the service account token that is passed to the credential provider plugin.

Only the `matchImages` parameter is configurable by using the `CRIOCredentialProviderConfig` object. All other parameters are immutable.

> In this example, the `crio-credential-provider` configuration was generated from the the `CRIOCredentialProviderConfig` object. The `gcr-credential-provider` configuration is a default configuration. 

Reviewer: one "the" can be removed

  name: cluster
spec:
  matchImages:
    - "docker.io"
    - "*.example.io"
    - "quay.io"
    - "registry.example.com:5000"
----
where:
+
--
`spec.matchImages`:: Specifies the registries that you want pods to pull from by using a project-scoped secrets. Each entry in the list must be a valid fully-qualified domain name with an optional wildcard, port, and path. The maximum length is 512 characters. Wildcards ('*') are supported for full-subdomain labels and top-level domains. Wildcards are not allowed in the port or path portions. The items in this list trigger the CRI-O credential provider. 
--
+
After you save the changes, {product-title} updates the `CredentialProviderConfig` object on the nodes. 

.Verification

. Verify that the `CredentialProviderConfig` object contains the registries from the `matchImages` list:

.. Start a debug session as root for a control plane node by running the following command:
+
[source,terminal]
----
$ oc debug node/<node_name>
----

.. Set `/host` as the root directory within the debug shell:
+
[source,terminal]
----
sh-5.1# chroot /host
----

.. View the `/etc/kubernetes/credential-providers/<platform>-credential-provider.yaml` file:
+

> [source,termial]

Reviewer: It should be [source, terminal]

----
sh-5.1# cat /etc/kubernetes/credential-providers/<platform>-credential-provider.yaml 

> ----

Reviewer: Could we add a brief note here or at the start about what the expected output should look like or atleast what the user should look for?

Author: Yeah, we have an [example CredentialProviderConfig above](https://github.com/openshift/openshift-docs/pull/111160/changes/BASE..01b51e673d041599f8f709b4ae5f1fd92dfaaab3#diff-37af2ca740e3707a529cf95f72c00669a25396b188c1cf6a3b82d08d57353dd9R46). I cannot decide whether the example is better there or here. Or, perhaps both place? WDYT?

Reviewer: I think we can keep the existing example above for context on what the object looks like, and add a short snippet here in the verification section for better visibility.

// Module included in the following assemblies:
//
// * openshift_images/image-configuration.adoc

:_mod-docs-content-type: PROCEDURE
[id="images-configuration-registry-mirror-project-secret_{context}"]
= Configuring project-scoped image pull secrets for mirrored registries

[role="_abstract"]

> As a cluster administrator, you can configure a `CRIOCredentialProviderConfig` object to enable project-scoped image pull secrets that you can use with mirrored repositories. By using project-scoped secrets, you can maintain security isolation between projects without exposing credentials in a global pull secret.

Reviewer 2: `configure a CRIOCredentialProviderConfig object` might be misleading, `cluster` object is only editable, we can explicitly document that like:

```suggestion
As a cluster administrator, you can configure `criocredentialproviderconfig.config.openshift.io/cluster` object to enable project-scoped image pull secrets that you can use with mirrored repositories. By using project-scoped secrets, you can maintain security isolation between projects without exposing credentials in a global pull secret.
```

> By default, if your cluster uses an `ImageDigestMirrorSet`, `ImageTagMirrorSet`, or `ImageContentSourcePolicy` object to configure repository mirroring, you must use a global pull secret for mirrored registries. You cannot add an image pull secret to a project. However, you can use a cluster-wide `CRIOCredentialProviderConfig` object to trigger the CRI-O credential provider, which enables project-scoped image pull secrets.

Reviewer 2: ```suggestion
By default, if your cluster uses an `ImageDigestMirrorSet`, `ImageTagMirrorSet`, or `ImageContentSourcePolicy` object to configure repository mirroring, you must use a global pull secret for mirrored registries. You cannot add an image pull secret to a project. However, you can use a cluster-wide `CRIOCredentialProviderConfig` object that configures kubelet to trigger the CRI-O credential provider, which enables project-scoped image pull secrets.
```

> This feature is supported for clusters on the {aws-first}, {gcp-first}, or {azure-first} platform.

Reviewer 2: we don't need to point out this, this feature also works for non cloud platform like vsphere, metal platforms. FYI, I updated the details of the enhancement https://github.com/openshift/enhancements/pull/1988

:FeatureName: Project-scoped image pull secrets for mirrored registries
include::snippets/technology-preview.adoc[]

> An administrator creates a cluster-wide `CRIOCredentialProviderConfig` object that lists the registries that a developer can pull from by using a project-scoped secrets. When a developer uses a pod spec to pull an image from a listed registry, the `CRIOCredentialProviderConfig` object triggers the CRI-O credential provider. The credential provider resolves mirror configurations, discovers namespace-scoped secrets, and generates short-lived authentication files for CRI-O consumption. This process maintains credential isolation between namespaces while preserving existing mirror configuration methods.

Reviewer 2: ```suggestion
An administrator edits cluster-wide `criocredentialproviderconfig.config.openshift.io/cluster` object that lists the registries that a developer can pull from by using a project-scoped secrets. When a developer uses a pod spec to pull an image from a listed registry, the `CRIOCredentialProviderConfig` object triggers the CRI-O credential provider. The credential provider resolves mirror configurations, discovers namespace-scoped secrets, and generates short-lived authentication files for CRI-O consumption. This process maintains credential isolation between namespaces while preserving existing mirror configuration methods.
```

> The administrator creates a image pull secret in each namespace where it is needed and configures role-based access control (RBAC) permissions to allow the pod to access the secret. The admin can create a different pull secret with different credentials for each namespace or use the same pull secret in multiple namespaces. 

Reviewer 2: ```suggestion
The administrator creates an image pull secret in each namespace where it is needed and configures role-based access control (RBAC) permissions to allow the pod's service account within that namespace to access the secret. The admin can create a different pull secret with different credentials for each namespace or use the same pull secret in multiple namespaces. 
```

The following is an example `CRIOCredentialProviderConfig` object:

[source,terminal]
----
apiVersion: config.openshift.io/v1alpha1
kind: CRIOCredentialProviderConfig
metadata:
  name: cluster
spec:
  matchImages:
    - "docker.io"
    - "*.example.io"
    - "quay.io"
    - "registry.example.com:5000"
----
where:

`providers.matchImages`:: Specifies the registries available for pulling by using project-scoped secrets. The items in this list trigger the CRI-O credential provider.

> Creating this object updates the `CredentialProviderConfig` object on the nodes.

Reviewer 2: ```suggestion
Editing this object updates the `CredentialProviderConfig` object on the nodes.
```

The following is an example `CredentialProviderConfig` object:

[source,terminal]
----
apiVersion: kubelet.config.k8s.io/v1
kind: CredentialProviderConfig
providers:
- apiVersion: credentialprovider.kubelet.k8s.io/v1
  args:
  - get-credentials
  - --v=3
  defaultCacheDuration: 1m0s
  matchImages:
  - gcr.io
  - '*.gcr.io'
  - '*.pkg.dev'
  - container.cloud.google.com
  name: gcr-credential-provider
- apiVersion: credentialprovider.kubelet.k8s.io/v1
  defaultCacheDuration: 1s
  matchImages:
  - '*.example.io'
  - docker.io
  - quay.io
  - registry.example.com:5000
  name: crio-credential-provider
  tokenAttributes:
    cacheType: Token
    requireServiceAccount: false
    serviceAccountTokenAudience: https://kubernetes.default.svc
----
where:

> `providers.defaultCacheDuration`:: Specifies the default duration the plugin will cache credentials in-memory. The default is `1m0s`.

Reviewer 2: This value is not in use. This field is required by the syntax, so this is a placeholder value provided to satisfy API schema requirements, it has no functional effect.

Only the `matchImages` parameter is configurable by using the `CRIOCredentialProviderConfig` object. All other parameters are immutable.

> In this example, the `crio-credential-provider` configuration was generated from the `CRIOCredentialProviderConfig` object. The `gcr-credential-provider` configuration is a default configuration. 

Reviewer 3: Lot of information before getting to the procedure. This feels like REFERENCE module material here, and maybe even more so in DITA. I'm still learning how DITA is going to work but it might be worth asking someone like Jeana if they think this information would be better suited in a REFERENCE topic for when we migrate to DITA.

Author: @stevsmit All apologies. I meant to mention this. I agree that this is a lot of material for a procedure. But, many of the other modules in this assembly have too much material without splitting it out. I went for consistency. WDYT?

// The pull secret step 2 is taken from creating-pull-secret
.Procedure

. Create a namespace for the pod by running the following command:
+
[source,terminal]
----
$ oc create namespace <namespace_name>
----
+
Replace `<namespace_name>` with a name for your namespace.

. Create a secret in that namespace from an existing authentication file:
+
.. For Docker clients using `.docker/config.json`, run the following command:
+
[source,terminal]
----
$ oc create secret generic <pull_secret_name> \
    --from-file=.dockerconfigjson=<path/to/.docker/config.json> \
    --type=kubernetes.io/dockerconfigjson
----
+
Replace `<pull_secret_name>` with a name for your secret and `<path/to/.docker/config.json>` with the path to your Docker `config.json` file.

.. For Podman clients using `.config/containers/auth.json`, run the following command:
+
[source,terminal]
----
$ oc create secret generic <pull_secret_name> \
     --from-file=<path/to/.config/containers/auth.json> \
     --type=kubernetes.io/podmanconfigjson
----
+
Replace `<pull_secret_name>` with a name for your secret and `<path/to/.config/containers/auth.json>` with the path to your Podman `auth.json` file.

> . Create a `Role` and a `RoleBinding` object in the namespace to grant service account access to image pull secrets.

Reviewer 3: On line 99, before you introduce the sub-steps, you use a colon. Here, you use a period. I generally try to use periods before sub-steps but I as long as we're consistent  either is probably fine.

.. Create a YAML file similar to the following example:
+
.Example `CRIOCredentialProviderConfig` object
[source,terminal]
----
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: <role_name>
  namespace: <namespace_name>
rules:
- apiGroups: [""]
  resources: ["secrets"]
  verbs: ["get", "list"]
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: credential-provider-secret-access-binding
  namespace: <namespace_name>
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: <role_name>
subjects:
- apiGroup: rbac.authorization.k8s.io
  kind: User
  name: system:serviceaccount:<namespace_name>:default
----
+
Replace `<namespace_name>` with a name for your namespace and `<role_name>` with a name for the role.

> .. Create the CR object:

Reviewer 3: maybe

```suggestion
.. Create the `CRIOCredentialProviderConfig` object:
```

+
[source,terminal]
----

> $ oc create -f <file-name>.yaml

Reviewer 3: ```suggestion
$ oc create -f <file_name>.yaml
```

----

.. Edit the `CRIOCredentialProviderConfig` object:
+
[source,terminal]
----
$ oc edit criocredentialproviderconfig cluster
----

.. Add the `spec.matchImages` stanza with a list of registries similar to the following example:
+
[source,terminal]
----
apiVersion: config.openshift.io/v1alpha1
kind: CRIOCredentialProviderConfig
metadata:
  name: cluster
spec:
  matchImages:
    - "docker.io"
    - "*.example.io"
    - "quay.io"
    - "registry.example.com:5000"
----
where:
+
--
`spec.matchImages`:: Specifies the registries that you want pods to pull from by using a project-scoped secret. Each entry in the list must be a valid fully-qualified domain name with an optional wildcard, port, and path. The maximum length is 512 characters. Wildcards ('*') are supported for full-subdomain labels and top-level domains. Wildcards are not allowed in the port or path portions. The items in this list trigger the CRI-O credential provider. 
--
+
After you save the changes, {product-title} updates the `CredentialProviderConfig` object on the nodes. 

.Verification

> . Verify that the `CredentialProviderConfig` object contains the registries from the `matchImages` list:

Reviewer 3: Same thought here as above.

Author: [ISG seems to suggest using a colon](https://www.ibm.com/docs/en/ibm-style?topic=format-procedures#writing-steps) before substeps. Though I didn't look all that hard:

1. Follow these steps to remove the drive:
a. Open the release levers on the left and right side of the drawer.
b. Slide the drawer out until it is fully extended.
c. Raise the DDM handle to release the drive that you want to replace.
d. Lift the drive out of the drawer.

. Verify that the `CredentialProviderConfig` object contains the registries from the `matchImages` list:

.. Start a debug session as root for a control plane node by running the following command:
+
[source,terminal]
----
$ oc debug node/<node_name>
----

.. Set `/host` as the root directory within the debug shell:
+
[source,terminal]
----
sh-5.1# chroot /host
----

.. View the `/etc/kubernetes/credential-providers/<platform>-credential-provider.yaml` file:
+
[source,terminal]
----
sh-5.1# cat /etc/kubernetes/credential-providers/<platform>-credential-provider.yaml 
----
+
Replace `<platform>` with one of the following options:
+
--
.. Use `ecr` for an {aws-short} cluster.
.. Use `gcr` for a {gcp-short} cluster.

> .. Use `acr` for an {azure-short} cluster.

Reviewer 2: we can explains that for non cloud platform, the CredentialProviderConfig file exists in `/etc/kubernetes/credential-providers/generic-credential-provider.yaml`

> .. Use `ecr` for an {aws-short} cluster.

Reviewer 3: This causes it to render as a, b, c, d. It's probably fine (though IIRC the IBMSG mentions something about that indicating sequential steps), and I wonder if ** might be better here.
